README.md
# VirusTotal-API
Scripts to query the VirusTotal API for URLs or file hashes.


Virus-Total-URL-Query-POSH/changelog.txt
Virus Total URL Query

--- Change Log ---
1.12
- Fixed an issue where queued URLs weren't being displayed in the output due to an incorrect for loop code section.

1.11
- Fixed a variable scope issue that was causing data not to be assigned to the correct scope.

1.10
- Added functionality to resubmit URLs for scanning if the scan_date is older than 6 months.
- Changed the way result data was displayed and exported.
    - If no malicious URLs were identified, then the results tables aren't displayed and nothing is exported.

1.05
- Added functionality to handle queued scans, which will be placed in the exported CSV file with permalink to the scan report.
- Added functionality to interrupt the scanning function and still continue with the results and export functions.

1.00
- Initial project version


Virus-Total-URL-Query-POSH/vt_url_query.ps1
# Virus Total URL Query script
# Version 1.12
 
# $api_key_file = Get-Content $script_path\api_key.txt | Out-String | ConvertFrom-StringData
$api_key = "" # $api_key_file.api_key
$start = Get-Date
$stop = Get-Date

function get_dir {
    $invoke = (Get-Variable MyInvocation -Scope 1).Value
    Split-Path $invoke.MyCommand.Path
}

$script_path = get_dir

$input_file = $script_path + "\domains.txt"
$cnt = 0

$mal_hits = 0
$mal_urls = [ordered]@{}

$queued_urls = New-Object System.Collections.Generic.List[System.Object]

 
function pause {
    Read-Host "`n`nPress Enter to continue." | Out-Null
}
 
function check_api_key {
    # $api_check = Get-Content $script_path\api_key.txt | Out-String | ConvertFrom-StringData
    If (-not $Script:api_key) {
        Write-Host "`n`n*** Virus Total API key not present in the api_key variable. ***`n" -ForegroundColor Yellow
        pause
        Break
    }
}
 
function welcome {
    Write-Host "`n`n================================`n|    Virus Total URL Query     |`n================================"
    Write-Host "`nInfo: Due to restrictions for public API use, `nInfo: there is a 26 second pause between queries.`n"
    Write-Host "================================`n"
}
 
function vt_search {
    $Script:start = Get-Date
    $estimate_time = $Script:start + (New-TimeSpan -Seconds (((Get-Content $Script:input_file | Measure-Object -Line).Lines - 1) * 26))
    $estimate_elapsed = $estimate_time - $Script:start
    
    Write-Host "Domains:" (Get-Content $Script:input_file | Measure-Object -Line).Lines
    Write-Host "Estimated time: $estimate_elapsed`n"
    Write-Host "Info: Press Ctrl-C to interrupt querying`n"
    Write-Host "Querying...`n--------------------------------"

    If ((Get-Content $Script:input_file | Measure-Object -Line).Lines -eq 0) {
        Write-Host "`n*** Domains.txt empty ***" -ForegroundColor Yellow
        break
    }
    
    Get-Content $Script:input_file | ForEach-Object {
        $Script:cnt += 1
        $totalLines = (Get-Content $Script:input_file | Measure-Object -Line).Lines
        
        Write-Host "$Script:cnt/$totalLines :: $_" -NoNewline
 
        $url = "http://www.virustotal.com/vtapi/v2/url/report?apikey=" + $Script:api_key + "&resource=" + $_
 
        $query = Invoke-RestMethod -Method "POST" -Uri $url
         
        If ($query.response_code -eq 0) {
            Write-Host "... No report. Scanning" -NoNewLine
            $query = vt_scan $_
        } ELSEIF ((($Script:start.ToUniversalTime()) - [datetime]$query.scan_date) -gt (New-TimeSpan -Days 180)) {
            Write-Host "... Old report. Scanning" -NoNewLine
            $query = vt_scan $_
        }
 
        If ($query.positives -gt 0) {
            $Script:mal_hits += 1
            $Script:mal_urls[$query.url] = [ordered]@{"url"=$query.url; "hits"=$query.positives; "scan_date"=$query.scan_date; "permalink"=$query.permalink}
            Write-Host (" ... Malicious {0}/{1}" -f $query.positives, $query.total) -ForegroundColor Red
        } ELSEIF ($query.response_code -eq 0) {
            $Script:mal_urls[$query.resource] = [ordered]@{"url"=$query.resource; "hits"=$null; "scan_date"="No scan report"; "permalink"="Perform manual scan of URL";}
            Write-Host (" ... No scan report") -ForegroundColor Yellow
        } ELSEIF (($query.verbose_msg -eq "Scan request successfully queued, come back later for the report") -OR ($query.response_code -eq "-2")) {
            Write-Host "... Queued. Try again later" -ForegroundColor Yellow
            $Script:queued_urls.Add($scan_url)
            $Script:mal_urls[$query.resource] = [ordered]@{"url"=$query.resource; "hits"=$null; "scan_date"="Queued"; "permalink"="Check later: "+$query.permalink;}
       } ELSE {
            Write-Host " ... Clean" -ForegroundColor Green
        }
        
        If ($Script:cnt -lt $totalLines) {
            Start-Sleep -Seconds 27
        }
    }
}
 
function vt_scan($scan_url) {
    $Local:url = "https://www.virustotal.com/vtapi/v2/url/scan?apikey=" + $Script:api_key + "&url=" + $scan_url
    
    $scan_query = Invoke-RestMethod -Method Post -Uri $Local:url
    
    return $scan_query
}
 
function show_results {
    $Script:stop = Get-Date
    
    Write-Host "`n================================"
    Write-Host "`n--- Results ---"
    Write-Host "`nPositive malicious hits: " -NoNewLine
    
    If ($Script:mal_hits -gt 0) {
        Write-Host "$Script:mal_hits" -ForegroundColor Red
           
        Write-Host "`n--------------------------------`n"
        Write-Host "List of malicious URLs:"
        Write-Host "`nHits `t URL`n--- `t ---"
    
        ForEach ($k in $Script:mal_urls.Keys) {
            If ($Script:mal_urls.$k.hits -gt 0) {
                Write-Host $Script:mal_urls.$k.hits `t $k
            }
        }

        Write-Host "`n--------------------------------`n"
        Write-Host "List of queued URLs`n`nInfo: Check permalink in CSV later"
        Write-Host "`nURL `t Verbose Msg `n--- `t -----------"

        $Script:queued_urls | ForEach-Object {
            Write-Host $_
        }
    
        Write-Host "`n--------------------------------"
        Write-Host "`nList of URLS without scan reports"
        Write-Host "`nURL `n---"
    
        ForEach ($k in $Script:mal_urls.Keys) {
            If ($Script:mal_urls.$k.scan_date -eq "No scan report") {
                Write-Host $Script:mal_urls.$k.url
            }
        }
    } Else {
        Write-Host "$Script:mal_hits" -ForegroundColor Green
    }
    
    Write-Host "`n================================`n"
}


function save_file {
    $date = Get-Date -UFormat %Y.%m.%d
    $time = Get-Date -UFormat %H.%M.%S
    $elapsed = $Script:stop - $Script:start
    
    Write-Host "--- Export Results ----"
    
    If ($Script:mal_urls.Count -gt 0) {
        Write-Host ("`n`Saving malicous urls to $date`_$time`_malicious_urls.csv")
        Write-Host ("Directory: $Script:script_path")
        ForEach ($k in $Script:mal_urls.Keys) {
            New-Object -typename psobject -property $Script:mal_urls.$k | Select-Object hits,url,scan_date,permalink | Export-Csv -Path "$Script:script_path\$date`_$time`_malicious_urls.csv" -NoTypeInformation -Append
        }
    } Else {
        Write-Host "`n*** Nothing to export ***" -ForegroundColor Yellow
    }
    
    Write-Host "`n`n--------------------------------"
    
    Write-Host ("`nStart: $Script:start `t Stop: $Script:stop `nElapsed: $elapsed")
}

welcome
check_api_key
Try {
    vt_search
}
Catch {
	$_ | Out-File ".\error.txt"
}
Finally {
    show_results
    save_file
    pause
}


Virus-Total-Hash-Query-POSH/generate_hashes_from_dir.ps1
# Generate Hashes from Directory script
# Place script file in directory with files to hash
# Version 0.1

function get_dir {
    $invoke = (Get-Variable MyInvocation -Scope 1).Value
    Split-Path $invoke.MyCommand.Path
}

$script_path = get_dir
$num_files = (Get-ChildItem -Recurse | Where-Object { -not $_.PSIsContainer }).Count - 1

$hashes = [ordered]@{}

function pause {
    Read-Host "`n`nPress Enter to continue." | Out-Null
}

function welcome {
    Write-Host "`n`n================================`n|      Generating Hashes       |`n================================"
    Write-Host "`nPath:  " -ForegroundColor Yellow -NoNewLine; Write-Host $script_path
    Write-Host "Files: " -ForegroundColor Yellow -NoNewLine; Write-Host $num_files"`n"
    Write-Host "================================`n"
}

function generate_hashes {
    $cnt = 0
    (Get-ChildItem -Recurse $script_path) | ForEach-Object {
        If (($_.Name -ne "generate_hashes_from_dir.ps1") -and (-not $_.PSIsContainer)) {
            $cnt += 1
            Set-Location $script_path
            $file_path = $_ | Resolve-Path -Relative
            $hash = (Get-FileHash -alg SHA1 -LiteralPath $file_path).Hash

            $hashes[$_] = [ordered]@{"filename"=$_.Name; "hash"=$hash; "alg"="SHA1"; "path"=$file_path;}
            Write-Host "$cnt... " -NoNewline
            If ($cnt -ge $num_files) {
                Write-Host "`n`n--------------------------------"
                Write-Host "`nFinished.`n" -ForegroundColor Green
            }
        }
    }

    Write-Host "================================`n"
}

function save_file {
    $date = Get-Date -UFormat %Y.%m.%d
    $time = Get-Date -UFormat %H.%M.%S
    
    Write-Host "--- Export Results ----"
    
    If ($hashes.Count -gt 0) {
        Write-Host ("`n`Saving hashes to $date`_$time`_hashes.csv")
        Write-Host ("Directory: $script_path")
        ForEach ($k in $hashes.Keys) {
            If ($k -ne "generate_hashes_from_dir.ps1") {
                New-Object -typename psobject -property $hashes.$k | Select filename,hash,alg,path | Export-Csv -Path "$Script:script_path\$date`_$time`_hashes.csv" -NoTypeInformation -Append
            }
        }
    } Else {
        Write-Host "`n*** Nothing to export ***" -ForegroundColor Yellow
    }
    
    Write-Host "`n`n--------------------------------"    
}

welcome
Try {
    generate_hashes
}
Catch {
	$_ | Out-File ".\error.txt"
}
Finally {
    save_file
    pause
}


Virus-Total-Hash-Query-POSH/vt_hash_query.ps1
# Virus Total Hash Query script
# Version 0.1
 
# $api_key_file = Get-Content $script_path\api_key.txt | Out-String | ConvertFrom-StringData
$api_key = "" # $api_key_file.api_key
$start = Get-Date
$stop = Get-Date

function get_dir {
    $invoke = (Get-Variable MyInvocation -Scope 1).Value
    Split-Path $invoke.MyCommand.Path
}

$script_path = get_dir

$input_file = Import-Csv $script_path"\hashes.csv"
$num_hashes = $input_file.Count
$cnt = 0

$mal_hits = 0
$mal_hashes = [ordered]@{}

$queued_urls = New-Object System.Collections.Generic.List[System.Object]

 
function pause {
    Read-Host "`n`nPress Enter to continue." | Out-Null
}
 
function check_api_key {
    # $api_check = Get-Content $script_path\api_key.txt | Out-String | ConvertFrom-StringData
    If (-not $Script:api_key) {
        Write-Host "`n`n*** Virus Total API key not present in the api_key variable. ***`n" -ForegroundColor Yellow
        pause
        Break
    }
}

function welcome {
    Write-Host "`n`n================================`n|    Virus Total Hash Query    |`n================================"
    Write-Host "`nInfo: Due to restrictions for public API use, `nInfo: there is a 26 second pause between queries.`n"
    Write-Host "================================`n"
}

function vt_search {
    $Script:start = Get-Date
    $estimate_time = $Script:start + (New-TimeSpan -Seconds ($num_hashes * 26))
    $estimate_elapsed = $estimate_time - $Script:start
    
    Write-Host "Hashes:" $num_hashes
    Write-Host "Estimated time: $estimate_elapsed`n"
    Write-Host "Info: Press Ctrl-C to interrupt querying`n"
    Write-Host "Querying...`n--------------------------------"

    If ($num_hashes -eq 0) {
        Write-Host "`n*** Hashes.csv empty ***" -ForegroundColor Yellow
        break
    }
    
    $Script:input_file | ForEach-Object {
        $Script:cnt += 1
        $totalLines = $num_hashes

        Write-Host "$Script:cnt/$totalLines :: $($_.filename)" -NoNewline
 
        $url = "http://www.virustotal.com/vtapi/v2/file/report?apikey=" + $Script:api_key + "&resource=" + $_.hash + "&allinfo=true"
 
        $query = Invoke-RestMethod -Method "GET" -Uri $url
 
        If ($query.positives -gt 0) {
            $Script:mal_hits += 1
            $Script:mal_hashes[$_.filename] = [ordered]@{"filename"=$_.filename; "hash"=$query.sha1; "path"=$_.path; "hits"=$query.positives; "scan_date"=$query.scan_date; "permalink"=$query.permalink}
            Write-Host (" ... Malicious {0}/{1}" -f $query.positives, $query.total) -ForegroundColor Red
        } ELSEIF ($query.response_code -eq 0) {
            $Script:mal_hashes[$_.filename] = [ordered]@{"filename"=$_.filename; "hash"=$_.hash; "path"=$_.path; "hits"=$null; "scan_date"="No scan report"; "permalink"="Perform local scan of file";}
            Write-Host (" ... No scan report.") -ForegroundColor Yellow
       } ELSE {
            Write-Host " ... Clean" -ForegroundColor Green
        }
        
        If ($Script:cnt -lt $totalLines) {
            Start-Sleep -Seconds 27
        }
    }
}
 
 
function show_results {
    $Script:stop = Get-Date
    
    Write-Host "`n================================"
    Write-Host "`n--- Results ---"
    Write-Host "`nPositive malicious hits: " -NoNewLine
    
    If ($Script:mal_hits -gt 0) {
        Write-Host "$Script:mal_hits" -ForegroundColor Red
        Write-Host "Please review the permalinks in the exported file." -ForegroundColor Yellow
           
        Write-Host "`n--------------------------------`n"
        Write-Host "List of malicious hashes:"
        Write-Host "`nHits `t Filename`n--- `t ---"
	} Else {
        Write-Host "$Script:mal_hits" -ForegroundColor Green
    }
    
	ForEach ($k in $Script:mal_hashes.Keys) {
		If ($Script:mal_hashes.$k.hits -gt 0) {
			Write-Host $Script:mal_hashes.$k.hits `t`t $Script:mal_hashes.$k.filename
		}
	}


	Write-Host "`n--------------------------------"
	Write-Host "`nList of hashes without scan reports:`n"

	ForEach ($k in $Script:mal_hashes.Keys) {
		If ($Script:mal_hashes.$k.scan_date -eq "No scan report") {
			Write-Host $Script:mal_hashes.$k.hash `t $Script:mal_hashes.$k.filename
		}
	}
    
    
    Write-Host "`n================================`n"
}


function save_file {
    $date = Get-Date -UFormat %Y.%m.%d
    $time = Get-Date -UFormat %H.%M.%S
    $elapsed = $Script:stop - $Script:start
    
    Write-Host "--- Export Results ----"
    
    If ($Script:mal_hashes.Count -gt 0) {
        Write-Host ("`n`Saving malicous hashes to $date`_$time`_malicious_hashes.csv")
        Write-Host ("Directory: $Script:script_path")
        ForEach ($k in $Script:mal_hashes.Keys) {
            New-Object -typename psobject -property $Script:mal_hashes.$k | Select hits,hash,filename,path,scan_date,permalink | Export-Csv -Path "$Script:script_path\$date`_$time`_malicious_hashes.csv" -NoTypeInformation -Append
        }
    } Else {
        Write-Host "`n*** Nothing to export ***" -ForegroundColor Yellow
    }
    
    Write-Host "`n`n--------------------------------"
    
    Write-Host ("`nStart: $Script:start `t Stop: $Script:stop `nElapsed: $elapsed")
}

welcome
check_api_key
Try {
    vt_search
}
Catch {
	$_ | Out-File ".\error.txt"
}
Finally {
    show_results
    save_file
    pause
}
